File: backend/FluffitMember/src/main/java/com/ssafy/fluffitmember/messagequeue/KafkaConsumer.java
package com.ssafy.fluffitmember.messagequeue;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.ssafy.fluffitmember.member.entity.Member;
import com.ssafy.fluffitmember.member.repository.MemberRepository;
import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Slf4j
@RequiredArgsConstructor
public class KafkaConsumer {

    private final MemberRepository memberRepository;

    @KafkaListener(topics = "coin-update", groupId = "consumerGroupId")
    @Transactional
    public void updateCoin(String kafkaMessage) {
        log.info("Kafka Message: ->" + kafkaMessage);

        Map<Object, Object> map = new HashMap<>();
        ObjectMapper mapper = new ObjectMapper();
        try {
            map = mapper.readValue(kafkaMessage, new TypeReference<Map<Object, Object>>() {});
        } catch (JsonProcessingException ex) {
            ex.printStackTrace();
        }
        Optional<Member> findMember = memberRepository.findByMemberId((String) map.get("memberId"));

        if(findMember.isPresent()){
            Member member = findMember.get();
            member.updateCoin(member.getCoin() - (Integer)map.get("price"));
            memberRepository.save(member);
        }
    }

    @KafkaListener(topics = "battle-point-update", groupId = "consumerGroupId")
    @Transactional
    public void updateBattlePoint(String kafkaMessage) {
        log.info("Kafka Message: ->" + kafkaMessage);

        Map<Object, Object> map = new HashMap<>();
        ObjectMapper mapper = new ObjectMapper();
        try {
            map = mapper.readValue(kafkaMessage, new TypeReference<Map<Object, Object>>() {});
        } catch (JsonProcessingException ex) {
            ex.printStackTrace();
        }
        Optional<Member> findMember = memberRepository.findByMemberId((String) map.get("memberId"));

        if(findMember.isPresent()){
            Member member = findMember.get();
            int currentPoint = member.getBattlePoint();
            int battlePointChanges = (Integer)map.get("battlePointChanges");
            int updatePoint  = currentPoint + battlePointChanges;
            if(updatePoint < 0){
                member.updatePoint(0);
            }else{
                member.updatePoint(member.getBattlePoint() + (Integer)map.get("battlePointChanges"));
            }
            LocalDateTime now = LocalDateTime.now();
            member.updateBattleDate(now);
            memberRepository.save(member);
        }
    }
}

File: backend/FluffitFlupet/src/main/kotlin/com/ssafy/fluffitflupet/messagequeue/KafkaConsumer.kt
package com.ssafy.fluffitflupet.messagequeue

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import com.ssafy.fluffitflupet.repository.MemberFlupetRepository
import com.ssafy.fluffitflupet.service.FlupetService
import kotlinx.coroutines.*
import kotlinx.coroutines.reactor.awaitSingleOrNull
import org.slf4j.LoggerFactory
import org.springframework.kafka.annotation.KafkaListener
import org.springframework.stereotype.Service
import kotlin.coroutines.CoroutineContext

@Service
class KafkaConsumer(
    private val memberFlupetRepository: MemberFlupetRepository
): CoroutineScope {
    private val job = Job()
    override val coroutineContext: CoroutineContext
        get() = Dispatchers.Default + job

    private val log = LoggerFactory.getLogger(FlupetService::class.java)

    @KafkaListener(topics = ["steps-update"], groupId = "consumerGroupId")
    fun update(message: String) {
        log.info("카프카로부터 제대로 메세지를 받나?? [$message]")
        //var map2 = HashMap<Any, Any>()
        launch { stepUpdate(message) }
    }

    suspend fun stepUpdate(message: String) {
        val mapper = ObjectMapper()
        val map: Map<Any, Any> = try {
            mapper.readValue(message, object : TypeReference<Map<Any, Any>>() {}) //String형을 json타입으로 변환
        } catch (ex: Exception) {
            ex.printStackTrace()
            return
        }

        val userId = map["memberId"] as String
        val steps = map["steps"] as Long
        val mflupet = withContext(Dispatchers.IO){ memberFlupetRepository.findByMemberIdAndIsDeadIsFalse(userId).awaitSingleOrNull() }
        if(mflupet != null){
            mflupet.steps = steps
            withContext(Dispatchers.IO){ memberFlupetRepository.save(mflupet).awaitSingleOrNull() }
        }
    }
}